File: homework/ba-nguyen/lesson-8/find-minimum-in-rotated-sorted-array.ts
function findMin(numbs: number[]): number {
  let lo = 0;
  let hi = numbs.length - 1;

  if (numbs[lo] <= numbs[hi]) return numbs[lo];

  while (lo <= hi) {
    const mid = lo + Math.floor((hi - lo) / 2);

    if (numbs[mid] > numbs[mid + 1]) return numbs[mid + 1];

    if (numbs[mid] >= numbs[lo]) {
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }

  return -1;
}

File: homework/Long_Nguyen/lesson_8/FindMinimumInRotatedSortedArray.java
/**
 * Created by Long Nguyen
 */
public class FindMinimumInRotatedSortedArray {
    public int findMin(int[] nums) {
        if (nums.length == 0)
            return -1;
        if (nums.length == 1)
            return nums[0];
        if (nums.length == 2)
            return Math.min(nums[0], nums[1]);

        int left = 0;
        int right = nums.length - 1;
        if (nums[right] > nums[0])
            return nums[0];

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[mid + 1]) {
                return nums[mid + 1];
            }
            if (nums[mid] < nums[mid - 1]) {
                return nums[mid];
            }
            if (nums[mid] > nums[0]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}
